pkg/sos/client.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2018 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package sos
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/gorilla/mux"
+)
+
+// RegistersNecessaryPatterns registers all the neccesary patterns needed
+// to make a service becomes a SoS client.
+func RegistersNecessaryPatterns(router *mux.Router) {
+	router.HandleFunc("/ping", func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("pong"))
+	}).Methods("GET")
+}
+
+// RegisterServiceWithSos tries to register a service with SoS.
+// If an non-nil error is returned, the service needs to exit immediately.
+func RegisterServiceWithSos(service string, port uint) error {
+	return registerServiceWithSos(service, port, "http://localhost:"+PortNum)
+}
+
+func registerServiceWithSos(service string, port uint, sosServerURL string) error {
+	m := RegisterReqJson{service, port}
+	return makeRequestToServer("POST", sosServerURL+"/register", m)
+}
+
+// UnregisterServiceWithSos makes a request to SoS Server to unregister the service.
+// This function should be called before a service exit.
+func UnregisterServiceWithSos(service string) error {
+	return unregisterServiceWithSos(service, "http://localhost:"+PortNum)
+}
+
+func unregisterServiceWithSos(service string, sosServerURL string) error {
+	m := UnRegisterReqJson{service}
+	return makeRequestToServer("POST", sosServerURL+"/unregister", m)
+}
+
+func makeRequestToServer(reqType, url string, reqJson interface{}) error {
+	b, err := json.Marshal(reqJson)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(reqType, url, bytes.NewBuffer(b))
+	if err != nil {
+		return err
+	}
+
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		decoder := json.NewDecoder(res.Body)
+		defer res.Body.Close()
+		var retMsg struct{ Error string }
+		if err := decoder.Decode(&retMsg); err != nil {
+			return err
+		}
+		if retMsg.Error != "" {
+			return fmt.Errorf(retMsg.Error)
+		}
+	}
+
+	return nil
+}
+
+// GetListener starts listener on a random port in localhost
+// and returns the listener and the port that the listener is on.
+// Remember to close the listener with:
+//
+// defer listener.close()
+//
+// or if it's used in a server, remeber to shutdown the server.
+func GetListener() (net.Listener, uint, error) {
+	listener, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return nil, 0, err
+	}
+
+	addrSplit := strings.Split(listener.Addr().String(), ":")
+	if len(addrSplit) != 2 {
```

Comment:
in the future, some person will thank you for this test.

Reply:
Fixed